New Data Source: aws_directory_service_directory

Tensho opened the pull request:
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):
```release-note
**New Data Source:** `aws_directory_service_directory` (#11282)
```

### Example

```hcl
resource "aws_directory_service_directory" "example" {
  ...
}

data "aws_directory_service_directory" "example" {
  directory_id = "aws_directory_service_directory.example.id"
}
```

### Acceptance Tests

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccDataSourceAwsDirectoryServiceDirectory'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccDataSourceAwsDirectoryServiceDirectory -timeout 120m
=== RUN   TestAccDataSourceAwsDirectoryServiceDirectory_SimpleAD
=== PAUSE TestAccDataSourceAwsDirectoryServiceDirectory_SimpleAD
=== RUN   TestAccDataSourceAwsDirectoryServiceDirectory_MicrosoftAD
=== PAUSE TestAccDataSourceAwsDirectoryServiceDirectory_MicrosoftAD
=== CONT  TestAccDataSourceAwsDirectoryServiceDirectory_SimpleAD
=== CONT  TestAccDataSourceAwsDirectoryServiceDirectory_MicrosoftAD
--- FAIL: TestAccDataSourceAwsDirectoryServiceDirectory_SimpleAD (755.12s)
    testing.go:635: Step 0 error: Check failed: 1 error occurred:
        	* Check 6/14 error: data.aws_directory_service_directory.test-simple-ad: Attribute 'password' not found


--- FAIL: TestAccDataSourceAwsDirectoryServiceDirectory_MicrosoftAD (1966.92s)
    testing.go:635: Step 0 error: Check failed: 1 error occurred:
        	* Check 6/14 error: data.aws_directory_service_directory.test-microsoft-ad: Attribute 'password' not found


FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	1968.212s
FAIL
make: *** [testacc] Error 1
```

### References

- [DescribeDirectories](https://docs.aws.amazon.com/directoryservice/latest/devguide/API_DescribeDirectories.html)

Tensho commented:
@bflad Please could you guide me why `password` attribute is not found in tests?
```
$ make testacc TEST=./aws TESTARGS='-run=TestAccDataSourceAwsDirectoryServiceDirectory'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccDataSourceAwsDirectoryServiceDirectory -timeout 120m
=== RUN   TestAccDataSourceAwsDirectoryServiceDirectory_SimpleAD
=== PAUSE TestAccDataSourceAwsDirectoryServiceDirectory_SimpleAD
=== RUN   TestAccDataSourceAwsDirectoryServiceDirectory_MicrosoftAD
=== PAUSE TestAccDataSourceAwsDirectoryServiceDirectory_MicrosoftAD
=== CONT  TestAccDataSourceAwsDirectoryServiceDirectory_SimpleAD
=== CONT  TestAccDataSourceAwsDirectoryServiceDirectory_MicrosoftAD
--- FAIL: TestAccDataSourceAwsDirectoryServiceDirectory_SimpleAD (755.12s)
    testing.go:635: Step 0 error: Check failed: 1 error occurred:
        	* Check 6/14 error: data.aws_directory_service_directory.test-simple-ad: Attribute 'password' not found


--- FAIL: TestAccDataSourceAwsDirectoryServiceDirectory_MicrosoftAD (1966.92s)
    testing.go:635: Step 0 error: Check failed: 1 error occurred:
        	* Check 6/14 error: data.aws_directory_service_directory.test-microsoft-ad: Attribute 'password' not found


FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	1968.212s
FAIL
make: *** [testacc] Error 1
```

Tensho commented:
@bflad Please could you guide me why `password` attribute is not found in tests?
```
$ make testacc TEST=./aws TESTARGS='-run=TestAccDataSourceAwsDirectoryServiceDirectory'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccDataSourceAwsDirectoryServiceDirectory -timeout 120m
=== RUN   TestAccDataSourceAwsDirectoryServiceDirectory_SimpleAD
=== PAUSE TestAccDataSourceAwsDirectoryServiceDirectory_SimpleAD
=== RUN   TestAccDataSourceAwsDirectoryServiceDirectory_MicrosoftAD
=== PAUSE TestAccDataSourceAwsDirectoryServiceDirectory_MicrosoftAD
=== CONT  TestAccDataSourceAwsDirectoryServiceDirectory_SimpleAD
=== CONT  TestAccDataSourceAwsDirectoryServiceDirectory_MicrosoftAD
--- FAIL: TestAccDataSourceAwsDirectoryServiceDirectory_SimpleAD (755.12s)
    testing.go:635: Step 0 error: Check failed: 1 error occurred:
        	* Check 6/14 error: data.aws_directory_service_directory.test-simple-ad: Attribute 'password' not found


--- FAIL: TestAccDataSourceAwsDirectoryServiceDirectory_MicrosoftAD (1966.92s)
    testing.go:635: Step 0 error: Check failed: 1 error occurred:
        	* Check 6/14 error: data.aws_directory_service_directory.test-microsoft-ad: Attribute 'password' not found


FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	1968.212s
FAIL
make: *** [testacc] Error 1
```

Tensho commented:
@gdavison Hi, could you help me with `password` attribute?

gdavison commented:
Hi @Tensho. The API doesn't return the password. That's definitely a good thing for security! :)

gdavison commented on aws/data_source_aws_directory_service_directory_test.go:
```diff
@@ -0,0 +1,162 @@
+
+func testAccDataSourceAwsDirectoryServiceDirectoryConfig_SimpleAD(alias string) string {
+	return fmt.Sprintf(`
+%s
```
Use `return fmt.Sprintf(...) + testAccDataSourceAwsDirectoryServiceDirectoryConfig_Prerequisites("simple-ad")`

gdavison commented on aws/data_source_aws_directory_service_directory_test.go:
```diff
@@ -0,0 +1,162 @@
+
+func testAccDataSourceAwsDirectoryServiceDirectoryConfig_MicrosoftAD(alias string) string {
+	return fmt.Sprintf(`
+%s
```
Use `return fmt.Sprintf(...) + testAccDataSourceAwsDirectoryServiceDirectoryConfig_Prerequisites("microsoft-ad")`

Tensho commented:
@gdavison Thank you, TBH, I thought sensitive attributes just returned back from the state on read as they were set there on create. Fixed typo `on-premis(e)`, removed `password` and replaced test config string interpolation with concatenation. Please let me know if I can do anything else here.

Tensho commented:
@gdavison Please let me know if I can do anything else here 🙇  I would be nice to squash all review commits on merge.
